### PR TITLE
refactor: code

### LINKED ISSUE #2121: caching failures with webpack serve when pointing to a specific config file
# Bug report

**What is the current behavior?**

If you change the name of your webpack configuration file and specify it with `webpack serve --config [filename]` you get caching errors in the console..

```
ℹ ｢wdm｣: Compiled with warnings.
<w> [webpack.cache.PackFileCacheStrategy] Caching failed for pack: Error: Can't resolve 'webpack.launcher.js' in '/Users/xxx/Projects/xxx/src/shell'
<w> while resolving 'webpack.launcher.js' in /Users/xxx/Projects/xxx/src/shell as file
```

**If the current behavior is a bug, please provide the steps to reproduce.**

Setup a webpack dev configuration where configuration X in extends from another configuration and merged though webpack merge.  (Though i don't think this is absolutely necessary).  Ensure that the configuration has file caching turned on.

```
  cache: {
    name: 'shell',
    type: 'filesystem',
    cacheDirectory: path.resolve(__dirname, '../..', '.cache'),
    buildDependencies: {
      config: [path.resolve(__dirname, 'webpack.common.js'), path.resolve(__filename)],
    },
  },
```

Launch `webpack serve --config webpack.launcher.js`.  This throw the above error. If however, you rename this file to webpack.config.js and run without --config.  It will work.

**What is the expected behavior?**

Should be able to name a webpack configuration file anything I want and point to it with --config when running `webpack serve` and filesystem caching should not be affected.

**Other relevant information:**
webpack version: 5.3.2
Node.js version: v12.19.0
Operating System: OSX
Additional tools:

### PR DESCRIPTION
**Pull request by alexander-akait**
**What kind of change does this PR introduce?**

bugfix, feature, refactoring, test

**Did you add tests for your changes?**

Yes

**If relevant, did you update the documentation?**

Maybe

**Summary**

- supports multi-compiler mode
- fix problems with cache
- improve performance
- refactor code
- `--no-watch`
- `--no-devtool`
- fix core flags and invalid values

**Does this PR introduce a breaking change?**

Potential yes, we remove resolving configuration based on `mode`, because it leads to unpredictable resolving, and remove ability to use the `mode` argument with default configuration, I want to mark it as bugfix and will provide information about it in changelog, but we should avoid to do it in future

**Other information**

fixes #2121

### CODE REVIEW

**Comment by snitin315** [comment]
CI is red. I will take care of it.

File: test/watch/watch-flag.test.js [hunk 1]
```diff
@@ -1,61 +1,42 @@
...
+    });
+
     it('should recompile upon file change', (done) => {
         const proc = runAndGetWatchProc(__dirname, ['--watch', '--mode', 'development'], false, '', true);
         let semaphore = 0;
         proc.stdout.on('data', (chunk) => {
             const data = stripAnsi(chunk.toString());
 
-            if (semaphore === 0 && data.includes('watching files for updates')) {
+            if (semaphore === 0 && data.includes('Compilation is watching files for updates...')) {
```
Comment (rishabh3112) [inline test/watch/watch-flag.test.js:26]:
```suggestion
            if (semaphore === 0 && data.includes('Compiler is watching files for updates...')) {
```

File: packages/webpack-cli/lib/plugins/CLIPlugin.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+            logger.log(`File '${filename}' was modified, changed time is ${date} (timestamp is ${changeTime})`);
+        });
+
+        (compiler.webpack ? compiler.hooks.afterDone : compiler.hooks.done).tap(pluginName, (stats) => {
+            logger.success(`Compilation${getCompilationName(stats.compilation)} finished`);
+
+            process.nextTick(() => {
+                if (compiler.watchMode) {
+                    logger.success(`Compilation${getCompilationName(stats.compilation)} is watching files for updates...`);
```
Comment (rishabh3112) [inline packages/webpack-cli/lib/plugins/CLIPlugin.js:78]:
```suggestion
                    logger.success(`Compiler${getCompilationName(stats.compilation)} is watching files for updates...`);
```
The compilation has finished already. Should be `Compiler` here as it calls instance of `Watching`.

**Review by rishabh3112 — COMMENTED** [review]
Code structure looks much better now. A minor suggestion.

File: test/watch/watch-flag.test.js (hunk 1 above)
Comment (alexander-akait) [reply test/watch/watch-flag.test.js:26]: Yes

File: packages/webpack-cli/lib/plugins/CLIPlugin.js (hunk 2 above)
Comment (alexander-akait) [reply packages/webpack-cli/lib/plugins/CLIPlugin.js:78]: Can you send a fix, I am working on the webpack-dev-server changelog, so we will do release with these fixes?

Comment (rishabh3112) [reply packages/webpack-cli/lib/plugins/CLIPlugin.js:78]: Did one at #2132

**Pull request merged**